### PR TITLE
Allow _ on LHS of assignment-stmt

### LINKED ISSUE #356: Allow `_`  on LHS of assignment
This means another alternative for lvexpr is `_`.

In the spec, this corresponds to the combination of

https://ballerina.io/spec/lang/2021R1/#destructuring-assignment-stmt
https://ballerina.io/spec/lang/2021R1/#wildcard-binding-pattern

This is important because it allows tests to avoid unused variables.

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fixes #356

### CODE REVIEW

**Comment by jclark** [comment]
Just treat _ as a keyword.

**Comment by KavinduZoysa** [comment]
@jclark, I have created DestructuringAssignStmt as follows.
```
public type DestructuringAssignStmt record {|
    BindingPattern pattern;
    Expr expr;
|};
```
When generating BIR, we have cast expr to any. Is that correct?

File: compiler/modules/front.syntax/ast.bal [hunk 1]
```diff
@@ -57,6 +57,14 @@ public type AssignStmt record {|
     Expr expr;
 |};
 
+const WildcardBindingPattern = "_";
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: This is more complicated than we need now. Just do

const WILDCARD = "_";

Then in AssignStmt

  LExpr|WILDCARD lValue;

File: compiler/modules/front.syntax/scan.bal [hunk 2]
```diff
@@ -133,7 +133,8 @@ final readonly & Keyword[] keywords = [
     "type",
     "typedesc",
     "while",
-    "xml"
+    "xml",
+    "_"
```
Comment (jclark) [inline compiler/modules/front.syntax/scan.bal]: Is this code point order?

File: compiler/modules/front.syntax/scan.bal [hunk 3]
```diff
@@ -567,7 +568,7 @@ function scanNormal(int[] codePoints, int startIndex, Scanned result) {
                     i = scanIdentifier(codePoints, i + 1, result);
                 }
                 else {
-                    endFragment(CP_UNDERSCORE, i, result);
+                    endIdentifierOrKeyword(codePoints, i, result);
```
Comment (jclark) [inline compiler/modules/front.syntax/scan.bal]: Just call endFragment here.  (You know what the keyword index is.)

File: compiler/modules/front.syntax/parseStmt.bal [hunk 4]
```diff
@@ -73,6 +73,9 @@ function parseStmt(Tokenizer tok) returns Stmt|err:Syntax {
         var td if td is InlineBuiltinTypeDesc|"map" => {
             return parseVarDeclStmt(tok);
         }
+        "_" => {
+            return parseDestructuringAssignStmt(tok);
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal]: Use finishAssignStmt

File: compiler/modules/front.syntax/parseStmt.bal [hunk 5]
```diff
@@ -222,14 +225,39 @@ function parseCompoundAssignStmt(Tokenizer tok, LExpr lValue, CompoundAssignOp o
...
+
 function parseVarDeclStmt(Tokenizer tok, boolean isFinal = false) returns VarDeclStmt|err:Syntax {
     TypeDesc td = check parseInlineTypeDesc(tok);
     return finishVarDeclStmt(tok, td, isFinal);
 }
 
 function finishVarDeclStmt(Tokenizer tok, TypeDesc td, boolean isFinal = false) returns VarDeclStmt|err:Syntax {
     Token? cur = tok.current();
-    string varName = check tok.expectIdentifier();
+    string varName;
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal]: This is not right. Do assignments first. Then do var decls as a separate pull request.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/parseStmt.bal [hunk 6]
```diff
@@ -222,14 +225,39 @@ function parseCompoundAssignStmt(Tokenizer tok, LExpr lValue, CompoundAssignOp o
...
+    return stmt;
+}
+
+function parseBindingPattern(Tokenizer tok) returns BindingPattern|err:Syntax {
+    Token? cur = tok.current();
+    check tok.advance();
+    if (cur == "_") {
+        return WildcardBindingPattern;
+    }
+    return parseError(tok, "unsupported binding pattern"); // TODO: check the message
```
Comment (manuranga) [inline compiler/modules/front.syntax/parseStmt.bal]: I think `tok.expect("_");` is sufficient, we are not doing custom error messages.

File: compiler/modules/front.syntax/ast.bal (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/ast.bal]: Fixed

File: compiler/modules/front.syntax/scan.bal (hunk 2 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/scan.bal]: Fixed

File: compiler/modules/front.syntax/scan.bal (hunk 3 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/scan.bal]: Fixed

File: compiler/modules/front.syntax/parseStmt.bal (hunk 4 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/parseStmt.bal]: Fixed

File: compiler/modules/front.syntax/parseStmt.bal (hunk 5 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/parseStmt.bal]: Removed

File: compiler/modules/front.syntax/parseStmt.bal (hunk 6 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/parseStmt.bal]: This part is removed

File: compiler/modules/front.syntax/astString.bal [hunk 7]
```diff
@@ -117,7 +117,12 @@ function stmtToWords(Word[] w, Stmt stmt) {
         w.push(";");
     }
     else if stmt is AssignStmt {
-        exprToWords(w, stmt.lValue);
+        LExpr|WILDCARD lValue = stmt.lValue;
+        if lValue is WILDCARD {
+            w.push(WILDCARD);
+        } else {
```
Comment (jclark) [inline compiler/modules/front.syntax/astString.bal]: style

File: compiler/modules/front/codeGen.bal [hunk 8]
```diff
@@ -777,9 +777,23 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
...
         return codeGenAssignToMember(cx, startBlock, env, lValue, expr);
     }
+    else {
+        return codeGenDestructuringAssignStmt(cx, startBlock, env, stmt);
+    }
+}
+
+function codeGenDestructuringAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:AssignStmt stmt) returns CodeGenError|StmtEffect {
+    var { lValue, expr } = stmt;
+    if (lValue is s:WILDCARD) {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: style

But don't do this: call this function only if you have a wildcard

File: compiler/modules/front/codeGen.bal [hunk 9]
```diff
@@ -760,7 +760,7 @@ function codeGenPanicStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environm
 
 function codeGenVarDeclStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:VarDeclStmt stmt) returns CodeGenError|StmtEffect {
     var { varName, initExpr, td, isFinal } = stmt;
-    if lookup(varName, env) !== () {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Deal with var decl separately

File: compiler/modules/front/codeGen.bal [hunk 10]
```diff
@@ -777,9 +777,23 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
...
     }
+    else {
+        return codeGenDestructuringAssignStmt(cx, startBlock, env, stmt);
+    }
+}
+
+function codeGenDestructuringAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:AssignStmt stmt) returns CodeGenError|StmtEffect {
+    var { lValue, expr } = stmt;
+    if (lValue is s:WILDCARD) {
+        s:VarDeclStmt declStmt = {td: "any", varName: s:WILDCARD, initExpr: stmt.expr, isFinal: false};
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Don't do this. Factor out common code from codeGenVarDeclStmt/codeGenAssignToVar into a function and call that.

File: compiler/modules/front/codeGen.bal [hunk 11]
```diff
@@ -777,9 +777,23 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
...
-    else {
+    else if lValue is s:MemberAccessExpr {
         return codeGenAssignToMember(cx, startBlock, env, lValue, expr);
     }
+    else {
+        return codeGenDestructuringAssignStmt(cx, startBlock, env, stmt);
+    }
+}
+
+function codeGenDestructuringAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:AssignStmt stmt) returns CodeGenError|StmtEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: codeGenAssignToWildcard

File: compiler/modules/front/codeGen.bal [hunk 12]
```diff
@@ -777,9 +777,23 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
     if lValue is s:VarRefExpr {
         return codeGenAssignToVar(cx, startBlock, env, lValue.varName, expr);
     }
-    else {
+    else if lValue is s:MemberAccessExpr {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Don't make the wildcard the last one, so you get better type checking

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/astString.bal (hunk 7 above)
Comment (KavinduZoysa) [reply compiler/modules/front.syntax/astString.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 8 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 9 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Removed the check related to WILDCARD from `codeGenVarDeclStmt` and factored out common part to a separate method to call from `codeGenAssignToWildcard`

File: compiler/modules/front/codeGen.bal (hunk 11 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 12 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 10 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal [hunk 13]
```diff
@@ -759,29 +759,41 @@ function codeGenPanicStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environm
...
+    return codeGenVarDeclStmtAfterLookup(cx, startBlock, env, stmt);
 }
 
 function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:AssignStmt stmt) returns CodeGenError|StmtEffect {
     var { lValue, expr } = stmt;
     if lValue is s:VarRefExpr {
         return codeGenAssignToVar(cx, startBlock, env, lValue.varName, expr);
     }
+    else if lValue is s:WILDCARD {
+        return codeGenAssignToWildcard(cx, startBlock, env, stmt);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: You are not factoring out quite right. There is no need to create bindings.  Don't create a VarDeclStmt.  Create a register of type t:ANY and then call a function which does foldExpr/codeGenExpr and adds the AssignInsn.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 13 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal [hunk 14]
```diff
@@ -777,11 +775,19 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
...
     }
+    else if lValue is s:WILDCARD {
+        return codeGenAssignToWildcard(cx, startBlock, env, expr);
+    }
     else {
         return codeGenAssignToMember(cx, startBlock, env, lValue, expr);
     }
 }
 
+function codeGenAssignToWildcard(CodeGenContext cx, bir:BasicBlock block, Environment env, s:Expr expr) returns CodeGenError|StmtEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This doesn't ned to be a separate function

File: compiler/modules/front/codeGen.bal [hunk 15]
```diff
@@ -777,11 +775,19 @@ function codeGenAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
...
+    bir:Register result = cx.createRegister(t:ANY, "_");
+    return { block: check genAssignInsn(cx, env, result, block, expr) };
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Inline result

File: compiler/modules/front/codeGen.bal [hunk 16]
```diff
@@ -802,11 +808,14 @@ function codeGenAssignToVar(CodeGenContext cx, bir:BasicBlock startBlock, Enviro
...
     }
     s:Expr foldedExpr = check cx.foldExpr(env, expr, unnarrowedReg.semType);
-    var { result: operand, block: nextBlock } = check codeGenExpr(cx, startBlock, env, foldedExpr);
+    return { block: check genAssignInsn(cx, env, unnarrowedReg, startBlock, foldedExpr), assignments };
+}
 
-    bir:AssignInsn assign = { result: unnarrowedReg, operand };
-    nextBlock.insns.push(assign);
-    return { block: nextBlock, assignments };
+function genAssignInsn(CodeGenContext cx, Environment env, bir:Register result, bir:BasicBlock block, s:Expr expr) returns CodeGenError|bir:BasicBlock {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This name is not consistent with the naming conventions in the file.  I suggest codeGenAssign

File: compiler/modules/front/codeGen.bal [hunk 17]
```diff
@@ -802,11 +808,14 @@ function codeGenAssignToVar(CodeGenContext cx, bir:BasicBlock startBlock, Enviro
...
+    var { result: operand, block: nextBlock } = check codeGenExpr(cx, block, env, expr);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This needs to do folding.

File: compiler/testSuite/01-function/assign8-v.bal [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import ballerina/io;
+
+function foo(int[] arr, int i) returns int[] {
+    arr[i] = i + 10;
+    return arr;
+}
+
+public function main() {
+    int[] arr = [];
```
Comment (jclark) [inline compiler/testSuite/01-function/assign8-v.bal:10]: Need many more tests.  In particular, you need

_ = 1;

to check folding.

**Review by jclark — CHANGES_REQUESTED** [review]
Also need to add `e` tests for when RHS is not a subtype of any.

File: compiler/testSuite/01-function/assign8-v.bal (hunk 18 above)
Comment (KavinduZoysa) [reply compiler/testSuite/01-function/assign8-v.bal:10]: Added

File: compiler/modules/front/codeGen.bal (hunk 17 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Added

File: compiler/modules/front/codeGen.bal (hunk 16 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Renamed

File: compiler/modules/front/codeGen.bal (hunk 15 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 14 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Removed

File: compiler/modules/front/codeGen.bal [hunk 19]
```diff
@@ -801,12 +801,15 @@ function codeGenAssignToVar(CodeGenContext cx, bir:BasicBlock startBlock, Enviro
         unnarrowedReg = unnarrowedBinding.reg;
         assignments = [ unnarrowedReg.number ];
     }
-    s:Expr foldedExpr = check cx.foldExpr(env, expr, unnarrowedReg.semType);
-    var { result: operand, block: nextBlock } = check codeGenExpr(cx, startBlock, env, foldedExpr);
+    return { block: check codeGenAssign(cx, env, startBlock, unnarrowedReg, expr, unnarrowedReg.semType), assignments };
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Better to split this over multiple lines.

**Review by jclark — COMMENTED** [review]
This is much improved. I have a more thorough look when I get back to my desk.

File: compiler/modules/front/codeGen.bal (hunk 19 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

**Review by jclark — APPROVED** [review]
Looks good
